File: 1549-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.java
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        int left=0;
        Deque<Integer> max=new ArrayDeque<>();
        Deque<Integer> min=new ArrayDeque<>();
        int maxLength=0;
        for(int right=0;right<nums.length;right++){
            while(!max.isEmpty() && nums[max.peekLast()]<nums[right]){
               max.pollLast();
            }
            max.addLast(right);
            while(!min.isEmpty() && nums[min.peekLast()]>nums[right]){
                min.pollLast();
            }
            min.addLast(right);
            while(nums[max.peekFirst()]-nums[min.peekFirst()]>limit){
                if(max.peekFirst()==left){
                    max.pollFirst();
                }
                if(min.peekFirst()==left){
                    min.pollFirst();
                }
                left++;
            }
            maxLength=Math.max(maxLength,right-left+1);
        } 
        return maxLength;
    }
}

File: 1549-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        deque<int> maxi;
        deque<int> mini;
        int left=0;
        int maxLength=0;
        for(int right=0;right<nums.size();right++){
            while(!maxi.empty() && nums[maxi.back()]<nums[right]){
                maxi.pop_back();
            }
            maxi.push_back(right);
            while(!mini.empty() && nums[mini.back()]>nums[right]){
                mini.pop_back();
            }
            mini.push_back(right);
            while(nums[maxi.front()]-nums[mini.front()]>limit){
                if(maxi.front()==left){
                    maxi.pop_front();
                }
                if(mini.front()==left){
                    mini.pop_front();
                }
                left++;
            }
            maxLength=max(maxLength,right-left+1);
        }
        return maxLength;
    }
};